kernel/src/fs/procfs/pid/mem.rs

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use crate::{
+    fs::{
+        procfs::template::{FileOps, ProcFileBuilder},
+        utils::{Inode, InodeMode},
+    },
+    prelude::*,
+    Process,
+};
+
+/// Represents the inode at either `/proc/[pid]/mem` or `/proc/[pid]/task/[tid]/mem`.
+pub struct MemFileOps(Arc<Process>);
+
+impl MemFileOps {
+    pub fn new_inode(process_ref: Arc<Process>, parent: Weak<dyn Inode>) -> Arc<dyn Inode> {
+        ProcFileBuilder::new(Self(process_ref))
+            .parent(parent)
+            // Reference: <https://github.com/torvalds/linux/blob/0ff41df1cb268fc69e703a08a57ee14ae967d0ca/fs/proc/base.c#L3341>
+            .mode(InodeMode::from_bits_truncate(0o600))
+            .build()
+            .unwrap()
+    }
+}
+
+impl FileOps for MemFileOps {
+    fn read_at(&self, offset: usize, writer: &mut VmWriter) -> Result<usize> {
+        let guard = self.0.lock_root_vmar();
+        let vmar = guard.as_ref().ok_or(Error::new(Errno::ENOENT))?;
+
+        let mut reader = vmar.vm_space().reader(offset, writer.avail())?;
```

Comment:
The `reader` and `writer` methods of `VmSpace` will report error if they are called by any task other than the one that activates the `VmSpace`. In practice, this mean that only the _current_ task can only access the `VmSpace` of its own. But `/proc/<pid>/mem` may be accessed by a process other than `pid`. 

Luckily, this restriction can be removed as `VmSpace` also exposes the `activate` method. So theoretically, a task can switch to the page table of any other task. But we need to think carefully about the implications of doing so and extend the API carefully.